src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCuMatMult.java

```diff
@@ -0,0 +1,361 @@
+		}
+		else if(isM1Sparse && !isM2Sparse) {
+			// -------------------------------------------------------------------------------------
+			// sparse-dense matrix multiplication
+			// Step 1: Allocate output => dense format
+			getDenseMatrixOutputForGPUInstruction(ec, instName, outputName, outRLen, outCLen); 
+			
+			// Step 2: Get the handles to sparse/dense pointers for left, right and output
+			CSRPointer A = left.getGPUObject(gCtx).getJcudaSparseMatrixPtr();
+			Pointer B = getDensePointer(gCtx, right, instName);
+			Pointer C = getDensePointer(gCtx, output, instName);
+			
+			// Step 3: Invoke the kernel
+			sparseDenseMatMult(gCtx, instName, C, A, B, left.getNumRows(), left.getNumColumns(), 
+				right.getNumRows(), right.getNumColumns(), outRLen, outCLen, isLeftTransposed, isRightTransposed);
+			// -------------------------------------------------------------------------------------
+		}
+		else {
+			// -------------------------------------------------------------------------------------
+			// dense-dense matrix multiplication 
+			if(isM1Sparse && !isM2Sparse) {
+				LOG.debug("Potential OOM as conversion of sparse input to dense");
+			}
+			
+			// Step 1: Allocate output => dense format
+			getDenseMatrixOutputForGPUInstruction(ec, instName, outputName, outRLen, outCLen); 
+			
+			// Step 2: Get the handles to sparse/dense pointers for left, right and output
+			Pointer A = getDensePointer(gCtx, left, instName);
+			Pointer B = getDensePointer(gCtx, right, instName);
+			Pointer C = getDensePointer(gCtx, output, instName);
+			
+			// Step 3: Invoke the kernel
+			denseDenseMatMult(getCublasHandle(gCtx), instName, C, A, B, params);
+			// -------------------------------------------------------------------------------------
+		}
+		return output;
+	}
+	
+	private static void sparseDenseMatMult(GPUContext gCtx, String instName, Pointer C, CSRPointer A, Pointer B,
+			long leftNumRows, long leftNumColumns, long rightNumRows, long rightNumColumns, long outRLen, long outCLen,
+			boolean isLeftTransposed, boolean isRightTransposed) throws DMLRuntimeException {
+		// t(C) = t(B) %*% t(A)
+		Pointer output = null;
+		if(outRLen != 1 && outCLen != 1) {
+			output = gCtx.allocate(outRLen*outCLen*Sizeof.DOUBLE);
+		}
+		else {
+			// no transpose required for vector output
+			output = C;
+		}
+		CuMatMultParameters params = new CuMatMultParameters(
+				rightNumRows, rightNumColumns,  leftNumRows, leftNumColumns,  !isRightTransposed, !isLeftTransposed);
+		denseSparseMatMult(getCusparseHandle(gCtx), instName, output, B, A, params);
+		if(outRLen != 1 && outCLen != 1) {
+			// Transpose: C = t(output)
+			JCublas2.cublasDgeam(gCtx.getCublasHandle(), cublasOperation.CUBLAS_OP_T, cublasOperation.CUBLAS_OP_T, 
+					toInt(outCLen), toInt(outRLen), one(), output, 
+					toInt(outRLen), zero(), new Pointer(), toInt(outRLen), C, toInt(outCLen));
+			gCtx.cudaFreeHelper(output, true);
```

Comment:
Do you need to force this `cudaFreeHelper` call to use the eager evaluation. Maybe you could let the user override it with the configuration option instead?

Reply:
Please add some timers here so that we can tell from the advanced stats if this code path was taken.
Ideally I would have liked to encapsulate this `transpose` function into a helper function somewhere and invoke that. The timers could be inside that `transpose` function and the resulting code would be easier to read.

Reply:
Let's clean up other methods in a subsequent PR and focus only on matmult.